### PR TITLE
Add Git Synchronization

### LINKED ISSUE #24: git synchroniztion
The possibility of new synchroniztion methods was mentioned way back in 2015 https://plus.google.com/+Orgzly/posts/S2nBbxVibsB

Did anything become of this? Are there plans to add git synchronization support to orgzly?

### PR DESCRIPTION
**Pull request by IvanMalison**
Fixes #24

### CODE REVIEW

**Comment by IvanMalison** [comment]
This is just a work in progress for the time being, but I wanted to put this up just in case anyone wants to take a look at the approach I am taking.

**Comment by IvanMalison** [comment]
@nevenz This is getting pretty close, although it's not quite there. Do you want to take a look at what I did with the UI and see if it's in line with what you would like? I had to make some changes to the existing code to make file browsing work that you might want to take a look at.

**Comment by nevenz** [comment]
I tried it and for now just took a quick look a the code...

> This is getting pretty close, although it's not quite there.

Should cloning work already? I'm getting `InvalidRemoteException` (I've added some logging, so lines might not match):

```
09-28 02:01:46.111 21800 21800 D RepoFactory: 1#main: ensureRepositoryExists: git@bitbucket.org:nevenz/foo.git file:/storage/emulated/0/Download/git-sync-test
09-28 02:01:46.111 21800 21800 W System.err: org.eclipse.jgit.api.errors.InvalidRemoteException: Invalid remote: origin
09-28 02:01:46.111 21800 21800 W System.err:    at org.eclipse.jgit.api.CloneCommand.call(CloneCommand.java:140)
09-28 02:01:46.111 21800 21800 W System.err:    at com.orgzly.android.repos.RepoFactory.ensureRepositoryExists(RepoFactory.java:102)
```

Might be what you mean by "not quite there", but since I tried it, I might as well report.

I get the same when I try to open the fragment to edit the repo (from the list of repositories). I don't think anything serious (like cloning) should happen at that point. BTW, all that code should probably be moved from `RepoFactory` to `GitRepo` itself.

Apart from few smaller UI stuff which I'm not going to mention now - when checkmark is clicked, repos are being created (looking at the logs), but fragment is not closed. Also, it would be helpful displaying some error messages (even in this early stage) if repo is not created due to missing data or whatever.

> Do you want to take a look at what I did with the UI and see if it's in line with what you would like? I had to make some changes to the existing code to make file browsing work that you might want to take a look at.

I like it, `FileSelectionFragment` could be used for `DirectoryRepo` too, or anything else in the future.

I can't select the key though. I had to select a directory, then type the file name.

**Comment by IvanMalison** [comment]
> I can't select the key though. I had to select a directory, then type the file name.

Oh oops, I guess maybe we should make it so that the browser can select a file?

> Should cloning work already?

No, not quite yet, I need to fiddle with where it actually puts the stuff, and how it provides credentials

> BTW, all that code should probably be moved from RepoFactory to GitRepo itself.

Yeah, actually I plan on doing the clone when the repo is created, to validate that all of the settings work as well, so I want to share the code somehow. I agree that it should be moved out of factory though.

> Also, it would be helpful displaying some error messages (even in this early stage) if repo is not created due to missing data or whatever.

Yeah agreed, thats another thing I need to get to.



I mostly just wanted to get feedback on whether the changes to the starting of the browser okay.

My main concern (and the reason I wanted to have you look at the UI stuff) is actually with the BrowserResultHandler browserResultHandler instance variable. I haven't really done any android development before this, but my understanding is that fragments can be destroyed at any time by the runtime if they aren't displayed. Could something weird happen here where that variable points to a dead fragment or something like that? I'm not 100% clear on the conditions under which fragments/activity can be destroyed/created.

**Comment by nevenz** [comment]
> Oh oops, I guess maybe we should make it so that the browser can select a file?

My idea with browser is to be reused for something like notes browsing too. For example for refiling when you have to choose the target note. So there could be implementation of `BrowserFragment` for that too.

Now, we either rename `FileBrowserFragment` to `DirectoryBrowserFragment` and create new `FileBrowserFragment` that would call `onBrowserUse` in `onListItemClick`, or improve current implementation to accept option whether to allow selecting files or not.

It's probably easier to just create a new implementation which would inherit the existing one, for now?

> My main concern (and the reason I wanted to have you look at the UI stuff) is actually with the BrowserResultHandler browserResultHandler instance variable.

It should be OK. *Activity* can killed by the system if it's not displayed, but as long as there is a fragment displayed and others are in the backstack, they should be fine.

BTW, looking at all those interfaces/listeners (and more coming in the future), I'm thinking we should switch to using activities per repo at one point (probably something I should have done from the start). It's getting confusing now who calls what etc. Maybe after this is released, I'll take a shot at that and see if it's worth it.

**Comment by IvanMalison** [comment]
> implementation to accept option whether to allow selecting files or not.

Yeah that seems like the most reasonable approach to me.

> using activities per repo at one point

Yes, definitely. The current setup is way too complicated.

@nevenz It would really make the rest of the work for this pull request a lot easier if we could move to storing repository type in the database instead of relying on uris to determine repo type. What is the advantage of doing things that way?

**Comment by nevenz** [comment]
> It would really make the rest of the work for this pull request a lot easier if we could move to storing repository type in the database instead of relying on uris to determine repo type.

Sounds good to me. Are you suggesting we do this *before* this PR, or as a part of it?

Which places besides `getFromUri` would that make things easier BTW?

**Comment by IvanMalison** [comment]
@nevenz git synchronization should be able to handle the case where there are both local and remote modifications. How would you refactor things to allow this?

**Comment by IvanMalison** [comment]
@nevenz @tulth I would like to pursue a much more aggresive refactor of syncing that gives a bit more control to the particular sync types implementation. The current implementation makes quite a few assumptions that seem tightly coupled to current implementations.

The idea would be to have the current code in Shelf.java to be one of multiple sync "strategies" that could be used. I think that at the very least, it should be up to the repo implementation to decide what the status of the sync is. The fact that this is not abstract causes a lot of issues.

File: app/src/main/java/com/orgzly/android/Shelf.java [hunk 1]
```diff
@@ -449,6 +449,17 @@ public BookAction syncNamesake(final BookNamesake namesake) throws IOException {
         String fileName;
         BookAction bookAction = null;
 
+        // XXX: This is a pretty nasty hack that completely circumvents the existing code path
+        VersionedRook rook = namesake.getLatestLinkedRook();
+        Repo repo = getRepo(rook.getRepoUri());
+        Repo.TwoWaySync sync = repo.getSync();
+        if (sync != null) {
+            handleTwoWaySync(sync, namesake);
```
Comment (IvanMalison) [inline app/src/main/java/com/orgzly/android/Shelf.java]: @nevenz I expect that this is probably not the way you would like me to handle this, but I need an interface like this to handle git synchronization properly.

Do you have ideas/feelings about how this should be done?

File: app/src/main/java/com/orgzly/android/Shelf.java [hunk 2]
```diff
@@ -449,6 +449,17 @@ public BookAction syncNamesake(final BookNamesake namesake) throws IOException {
...
+            return new BookAction(
+                    BookAction.Type.INFO,
+                    namesake.getStatus().msg(UriUtils.friendlyUri(repo.getUri().toString())));
+        }
+
         switch (namesake.getStatus()) {
             case NO_CHANGE:
```
Comment (IvanMalison) [inline app/src/main/java/com/orgzly/android/Shelf.java]: I feel like all of this should not be required of all synchronization strategies i.e. it should belong in its own class that CAN be used for some of the repos, but not others.

File: app/src/main/java/com/orgzly/android/repos/Repo.java [hunk 3]
```diff
@@ -42,4 +42,12 @@
     void delete(Uri uri) throws IOException;
 
     String toString();
+
+    TwoWaySync getSync();
```
Comment (IvanMalison) [inline app/src/main/java/com/orgzly/android/repos/Repo.java]: This is obviously hideous, I have a few ideas about how to get around it, but they require a lot of changes in Shelf.java

**Comment by IvanMalison** [comment]
Synchronization sort of works now if anyone would like to try this. Some things (ssh key selection, directory naming, overlapping filenames across repositories) might be weird.

**Comment by jgkamat** [comment]
Hi! I'm trying to test this out, and I'm having a bit of a problem building it, and I'm not exactly sure whats going on. Could someone take a look at my build? The [full output is here](http://paste.debian.net/988714/). 

Otherwise, can someone provide me with an APK so I can try this out! I'm super excited for this, I've been waiting for it for a long time!

**Comment by IvanMalison** [comment]
@jgkamat Not sure what is going on there. Here's an apk: https://nofile.io/f/tdSBOAeOzS1/orgzly-git-sync.apk

**Comment by jgkamat** [comment]
Im getting a crash on startup with the version you gave, the version from fdroid seems fine.

```
I/ActivityManager( 5140): START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10300000 pkg=com.orgzly cmp=com.orgzly/.android.ui.MainActivity} from uid 1000 on display 0
I/ActivityManager( 5140): Start proc 29047:com.orgzly/u0a609 for activity com.orgzly/.android.ui.MainActivity
W/System  (29047): ClassLoader referenced unknown path: /data/app/com.orgzly-1/lib/arm
D/com.orgzly.android.provider.Provider(29047): 1#main: onCreate
D/com.orgzly.android.provider.Database(29047): 1#main: <init>: orgzly.db 143
D/com.orgzly.android.ui.MainActivity(29047): 1#main: onCreate: null
D/com.orgzly.android.ui.CommonActivity(29047): 1#main: onCreate: null
D/com.orgzly.android.ui.MainActivity(29047): 1#main: setupDrawer
D/com.orgzly.android.ui.MainActivity(29047): 1#main: addSyncFragment
D/com.orgzly.android.ui.fragments.SyncFragment(29047): 1#main: <init>
D/com.orgzly.android.ui.fragments.DrawerFragment(29047): 1#main: <init>
D/com.orgzly.android.ui.MainActivity(29047): 1#main: setupDisplay: null
D/com.orgzly.android.ui.DisplayManager(29047): 1#main: isFragmentDisplayed: tag:com.orgzly.android.ui.fragments.BooksFragment fragment:null isVisible:no
D/com.orgzly.android.ui.fragments.BooksFragment(29047): 1#main: <init>
D/com.orgzly.android.ui.fragments.SyncFragment(29047): 1#main: onAttach: com.orgzly.android.ui.MainActivity@e88da65
D/com.orgzly.android.ui.MainActivity(29047): 1#main: onAttachFragment: SyncFragment{fa7f01c #0 id=0x7f1000ac com.orgzly.android.ui.fragments.SyncFragment}
D/com.orgzly.android.ui.fragments.SyncFragment(29047): 1#main: onCreate: null
D/com.orgzly.android.ui.fragments.SyncFragment(29047): 1#main: onCreateView: com.android.internal.policy.PhoneLayoutInflater@fe669fa android.widget.FrameLayout{ac17bab V.E...... ......I. 0,0-0,0 #7f1000ac app:id/drawer_sync_container} null
D/com.orgzly.android.ui.fragments.DrawerFragment(29047): 1#main: onAttach: com.orgzly.android.ui.MainActivity@e88da65
D/com.orgzly.android.ui.MainActivity(29047): 1#main: onAttachFragment: DrawerFragment{1306905 #1 id=0x7f1000ab com.orgzly.android.ui.fragments.DrawerFragment}
D/com.orgzly.android.ui.fragments.DrawerFragment(29047): 1#main: onCreateView: com.android.internal.policy.PhoneLayoutInflater@7a6d5a android.widget.FrameLayout{f0e2b8b V.E...... ......I. 0,0-0,0 #7f1000ab app:id/drawer_list_container} null
D/com.orgzly.android.ui.fragments.DrawerFragment(29047): 1#main: onActivityCreated: null
D/com.orgzly.android.ui.fragments.DrawerFragment(29047): 1#main: onCreateLoader: 7 null
D/com.orgzly.android.ui.fragments.DrawerFragment(29047): 1#main: onCreateLoader: 6 null
D/com.orgzly.android.ui.fragments.BooksFragment(29047): 1#main: onAttach: com.orgzly.android.ui.MainActivity@e88da65
D/com.orgzly.android.ui.MainActivity(29047): 1#main: onAttachFragment: BooksFragment{ed94db2 #2 id=0x7f100079 com.orgzly.android.ui.fragments.BooksFragment}
D/com.orgzly.android.ui.fragments.BooksFragment(29047): 1#main: onCreate: null
D/com.orgzly.android.ui.fragments.BooksFragment(29047): 1#main: onCreateView: com.android.internal.policy.PhoneLayoutInflater@1de8103 android.widget.FrameLayout{25e0880 V.E...... ......I. 0,0-0,0 #7f100079 app:id/single_pane_container} null
D/com.orgzly.android.ui.fragments.BooksFragment(29047): 1#main: onViewCreated: android.widget.RelativeLayout{94491b0 V.E...... ......ID 0,0-0,0 #7f1000b3 app:id/fragment_books_container} null
D/com.orgzly.android.ui.fragments.BooksFragment(29047): 1#main: onActivityCreated: null
D/com.orgzly.android.ui.fragments.BooksFragment(29047): 1#main: onCreateLoader: 2 null
D/com.orgzly.android.ui.fragments.BooksFragment(29047): 1#main: onViewStateRestored: null
D/com.orgzly.android.ui.MainActivity(29047): 1#main: onPostCreate: null
D/com.orgzly.android.ui.MainActivity(29047): 1#main: onResume
D/com.orgzly.android.ui.CommonActivity(29047): 1#main: onResume
D/com.orgzly.android.provider.Provider(29047): 4832#ModernAsyncTask #1: query: content://com.orgzly/books null null null is_dummy,LOWER(COALESCE(books_view.title, books_view.name))
D/com.orgzly.android.provider.Provider(29047): 4833#ModernAsyncTask #2: query: content://com.orgzly/filters null null null position, _id
D/com.orgzly.android.ui.MainActivity(29047): 1#main: onResumeFragments
D/com.orgzly.android.ui.fragments.DrawerFragment(29047): 1#main: onCreateLoader: 7 null
D/com.orgzly.android.ui.fragments.BooksFragment(29047): 1#main: onResume
D/com.orgzly.android.ui.fragments.BooksFragment(29047): 1#main: onCreateLoader: 2 null
D/com.orgzly.android.provider.Provider(29047): 4834#ModernAsyncTask #3: query: content://com.orgzly/books null null null is_dummy,LOWER(COALESCE(books_view.title, books_view.name))
D/com.orgzly.android.provider.Provider(29047): 4835#ModernAsyncTask #4: query: content://com.orgzly/books null null null is_dummy,LOWER(COALESCE(books_view.title, books_view.name))
D/com.orgzly.android.ui.MainActivity(29047): 1#main: announceChanges: com.orgzly.android.ui.fragments.BooksFragment Notebooks null 0
D/com.orgzly.android.provider.Provider(29047): 4836#ModernAsyncTask #5: query: content://com.orgzly/books null null null is_dummy,LOWER(COALESCE(books_view.title, books_view.name))
D/com.orgzly.android.provider.Provider(29047): 4834#ModernAsyncTask #3: query: Cursor count: 1 for books_view null null
D/com.orgzly.android.provider.Provider(29047): 4832#ModernAsyncTask #1: query: Cursor count: 1 for books_view null null
D/com.orgzly.android.provider.Provider(29047): 4833#ModernAsyncTask #2: query: Cursor count: 2 for searches null null
D/com.orgzly.android.provider.Provider(29047): 4836#ModernAsyncTask #5: query: Cursor count: 1 for books_view null null
D/com.orgzly.android.provider.Provider(29047): 4835#ModernAsyncTask #4: query: Cursor count: 1 for books_view null null
D/com.orgzly.android.sync.SyncService(29047): 1#main: onCreate
D/com.orgzly.android.sync.SyncService(29047): 1#main: onBind
D/com.orgzly.android.ui.MainActivity(29047): 1#main: onCreateOptionsMenu: android.support.v7.view.menu.MenuBuilder@ec3fa55
I/ActivityManager( 5140): Displayed com.orgzly/.android.ui.MainActivity: +423ms
D/com.orgzly.android.ui.fragments.BooksFragment(29047): 1#main: onCreateOptionsMenu: android.support.v7.view.menu.MenuBuilder@ec3fa55 android.support.v7.view.SupportMenuInflater@37922c5
D/com.orgzly.android.ui.fragments.DrawerFragment(29047): 1#main: onLoadFinished: CursorLoader{83447e6 id=6} android.content.ContentResolver$CursorWrapperInner@e0ee127
D/com.orgzly.android.ui.fragments.BooksFragment(29047): 1#main: onLoadFinished: CursorLoader{19efd4 id=2} android.content.ContentResolver$CursorWrapperInner@415e57d
D/com.orgzly.android.ui.fragments.DrawerFragment(29047): 1#main: onLoadFinished: CursorLoader{4ecfc72 id=7} android.content.ContentResolver$CursorWrapperInner@ffe68c3
E/AndroidRuntime(29047): Process: com.orgzly, PID: 29047
E/AndroidRuntime(29047): java.lang.NullPointerException: Attempt to invoke virtual method 'long com.orgzly.android.repos.VersionedRook.getMtime()' on a null object reference
E/AndroidRuntime(29047): 	at com.orgzly.android.Book.isModifiedAfterLastSync(Book.java:82)
E/AndroidRuntime(29047): 	at com.orgzly.android.ui.fragments.DrawerFragment.updateFromBooksCursor(DrawerFragment.java:362)
E/AndroidRuntime(29047): 	at com.orgzly.android.ui.fragments.DrawerFragment.onLoadFinished(DrawerFragment.java:307)
E/AndroidRuntime(29047): 	at com.orgzly.android.ui.fragments.DrawerFragment.onLoadFinished(DrawerFragment.java:38)
W/ActivityManager( 5140):   Force finishing activity com.orgzly/.android.ui.MainActivity
W/ActivityManager( 5140): Activity pause timeout for ActivityRecord{b354fc3 u0 com.orgzly/.android.ui.MainActivity t9558 f}
W/InputDispatcher( 5140): channel '9ae9917 com.orgzly/com.orgzly.android.ui.MainActivity (server)' ~ Consumer closed input channel or an error occurred.  events=0x9
E/InputDispatcher( 5140): channel '9ae9917 com.orgzly/com.orgzly.android.ui.MainActivity (server)' ~ Channel is unrecoverably broken and will be disposed!
I/WindowState( 5140): WIN DEATH: Window{9ae9917 u0 com.orgzly/com.orgzly.android.ui.MainActivity}
W/InputDispatcher( 5140): Attempted to unregister already unregistered input channel '9ae9917 com.orgzly/com.orgzly.android.ui.MainActivity (server)'
I/ActivityManager( 5140): Process com.orgzly (pid 29047) has died
W/ActivityManager( 5140): Scheduling restart of crashed service com.orgzly/com.evernote.android.job.JobRescheduleService in 1000ms
```

**Comment by IvanMalison** [comment]
@jgkamat Ahh oops, i know why thats happening.

**Comment by nevenz** [comment]
What about:

```
public interface TwoWaySyncable {
    VersionedRook syncBook(VersionedRook current, File fromDB, File writeBack) throws IOException;
}
```

leaving `Repo` as it was (could be renamed later), make only `GitRepo` implement `TwoWaySyncable` and in `syncNamesake` only have:

```
        if (repo instanceof TwoWaySyncable) {
            return handleTwoWaySync(repo, namesake);
        } else {
            return handleBookSync(repo, namesake);
        }
```

**Comment by IvanMalison** [comment]
> leaving Repo as it was (could be renamed later), make only GitRepo implement TwoWaySyncable and in syncNamesake only have:

@nevenz I'm definitely fine with that if it makes you happy.

**Comment by jgkamat** [comment]
I wasn't able to get past the repository creation screen, the checkbox didn't seem to let me beyond the settings page. In the logs, there were a bunch of errors relating to 'could not delete' the directory where the git repo was to be cloned. after deleting it manually, the error still persisted. 

Also, once trying out the "Directory" sync after trying that out, I got a bunch of errors, and the file didn't import properly. Switching to the release version worked fine, so I think that this PR might need some more work :cry:

**Comment by IvanMalison** [comment]
@nevenz Random question -- Did you ever consider letting the actual files be the canonical source of data for orgzly? Given how quickly orgzly was able to sync my 100+ file repository, it seems like parsing etc. is quick enough that everything could be done in memory. I have a hard time believing that peformance is a huge concern in this case -- this is, after all, a TODO application, and we are not dealing with millions of pieces of DATA.

**Comment by nevenz** [comment]
> Did you ever consider letting the actual files be the canonical source of data for orgzly? Given how quickly orgzly was able to sync my 100+ file repository, it seems like parsing etc. is quick enough that everything could be done in memory. I have a hard time believing that peformance is a huge concern in this case -- this is, after all, a TODO application, and we are not dealing with millions of pieces of DATA.

Briefly in the beginning. But I didn't see that working very well or fast, without re-inventing a lot of things that you get when you use DB.

Some of the reasons that come to mind...

You'd have to rewrite the whole file after every note modification. Doing a lot of small modifications to just a single large notebook (which sounds like something that it's done often, at least I do) could be a battery killer. It's a similar issue to the one auto-sync now has.

Search across all files would depend too much on number of files and their size and it could be very slow. In addition, it's way too easy to do a search with a complicated query now using a DB.

Also, org files are just one of the possible file formats Orgzly supports for exporting and importing data. Something like [BookName.Format](https://github.com/orgzly/orgzly-android/blob/89bc7454ccd76d127f4178856248659d77713d2e/app/src/main/java/com/orgzly/android/BookName.java#L92) is there from day 1 and is used throughout the code. I did get lazy over time, so there are quite a few Org-specific things in the code now. But it's one of the reasons I didn't consider using an org file as a way to store notes.

**Comment by IvanMalison** [comment]
@nevenz This is pretty much ready to go.

There are several serious bug when the uri of a git repo is edited, but it is not because of any new code.

**Comment by IvanMalison** [comment]
@jgkamat Sorry about that, I've fixed a bunch of the issues. There might be a few rough spots with this version, but it should at least pull in your repository.

**Comment by nevenz** [comment]
_Random list of things I noticed. It's a mix of issues and things I'd eventually do before including it in the release. There's some nitpicking going on too. I'll keep this comment updated._

- [ ] Warning when syncing from Android Studio
```
Warning:WARNING: Dependency org.apache.httpcomponents:httpclient:4.1.3 is ignored for fdroidDebug as it may be conflicting with the internal version provided by Android.
```
- [ ] Building fails when doing `./gradlew build` (`proguard-rules.pro` needs updating probably)
- [ ] Big button for Git needs to be added in repos list (between Dropbox and Directory)
- [ ] Change order in `+` drop-down (Dropbox, Git, Directory)
- [ ] Asking for Storage permission *before* entering Git repo creation, leaving user in repos list
- [ ] Can't scroll down (not using ScrollView?)
- [ ] Action button (`imeOptions`) - `actionNext` for all but last, which should be `actionDone`
- [ ] Remove "Git" from all hints
- [ ] Git author (e.g. My Name) -> Name (e.g. John Doe)
- [ ] Git email (e.g. me@email.com) -> Email (e.g. john@doe.com)
- [ ] Git branch (e.g. master) -> Branch
- [ ] Prepopulate branch with "master"
- [ ] Field with browse button don't have hint displayed when value is set (`TextInputLayout` missing?)
- [ ] Prepopulate directory (Orgzly/repos/git in storage root or similar)
- [ ] Git remote address -> Remote URL
- [ ] On tapping checkmark getting error: `IOException: Directory /storage/emulated/0/Download/git-sync-test is not a git repository`
- [ ] Extract string resources displayed to user (errors etc.)
- [ ] Appending "/non-existent" to directory crashes the app
```
10-12 23:20:20.008  6287  6416 E AndroidRuntime: Caused by: java.lang.ClassCastException: org.eclipse.jgit.transport.TransportHttp cannot be cast to org.eclipse.jgit.transport.SshTransport
10-12 23:20:20.008  6287  6416 E AndroidRuntime:        at com.orgzly.android.git.GitSSHKeyTransportSetter$2.configure(GitSSHKeyTransportSetter.java:40)
10-12 23:20:20.008  6287  6416 E AndroidRuntime:        at org.eclipse.jgit.api.TransportCommand.configure(TransportCommand.java:138)
10-12 23:20:20.008  6287  6416 E AndroidRuntime:        at org.eclipse.jgit.api.FetchCommand.call(FetchCommand.java:128)
10-12 23:20:20.008  6287  6416 E AndroidRuntime:        at org.eclipse.jgit.api.CloneCommand.fetch(CloneCommand.java:193)
10-12 23:20:20.008  6287  6416 E AndroidRuntime:        at org.eclipse.jgit.api.CloneCommand.call(CloneCommand.java:133)
10-12 23:20:20.008  6287  6416 E AndroidRuntime:        at com.orgzly.android.repos.GitRepo.ensureRepositoryExists(GitRepo.java:91)
10-12 23:20:20.008  6287  6416 E AndroidRuntime:        at com.orgzly.android.repos.GitRepo.ensureRepositoryExists(GitRepo.java:73)
10-12 23:20:20.008  6287  6416 E AndroidRuntime:        at com.orgzly.android.ui.fragments.GitRepoFragment$RepoCloneTask.doInBackground(GitRepoFragment.java:343)
10-12 23:20:20.008  6287  6416 E AndroidRuntime:        at com.orgzly.android.ui.fragments.GitRepoFragment$RepoCloneTask.doInBackground(GitRepoFragment.java:312)
10-12 23:20:20.008  6287  6416 E AndroidRuntime:        at android.os.AsyncTask$2.call(AsyncTask.java:305)
10-12 23:20:20.008  6287  6416 E AndroidRuntime:        at java.util.concurrent.FutureTask.run(FutureTask.java:237)
```
- [ ] Trying again, creating repo from scratch (somehow I didn't get IOException like above this time), sync fails, in logs:
```
10-12 23:37:19.958  6513  6589 W System.err: java.lang.NullPointerException: Attempt to read from field 'int org.eclipse.jgit.lib.AnyObjectId.w1' on a null object reference
10-12 23:37:19.959  6513  6589 W System.err:    at org.eclipse.jgit.lib.ObjectIdOwnerMap.get(ObjectIdOwnerMap.java:131)
10-12 23:37:19.959  6513  6589 W System.err:    at org.eclipse.jgit.revwalk.RevWalk.parseAny(RevWalk.java:838)
10-12 23:37:19.959  6513  6589 W System.err:    at org.eclipse.jgit.revwalk.RevWalk.parseCommit(RevWalk.java:753)
10-12 23:37:19.959  6513  6589 W System.err:    at com.orgzly.android.git.GitFileSynchronizer.getCommit(GitFileSynchronizer.java:240)
10-12 23:37:19.959  6513  6589 W System.err:    at com.orgzly.android.git.GitFileSynchronizer.currentHead(GitFileSynchronizer.java:233)
10-12 23:37:19.959  6513  6589 W System.err:    at com.orgzly.android.git.GitFileSynchronizer.setBranchAndGetLatest(GitFileSynchronizer.java:197)
10-12 23:37:19.959  6513  6589 W System.err:    at com.orgzly.android.repos.GitRepo.getBooks(GitRepo.java:186)
10-12 23:37:19.959  6513  6589 W System.err:    at com.orgzly.android.Shelf.getBooksFromAllRepos(Shelf.java:435)
10-12 23:37:19.959  6513  6589 W System.err:    at com.orgzly.android.Shelf.groupAllNotebooksByName(Shelf.java:400)
10-12 23:37:19.959  6513  6589 W System.err:    at com.orgzly.android.sync.SyncService$SyncTask.doInBackground(SyncService.java:243)
10-12 23:37:19.959  6513  6589 W System.err:    at com.orgzly.android.sync.SyncService$SyncTask.doInBackground(SyncService.java:226)
10-12 23:37:19.959  6513  6589 W System.err:    at android.os.AsyncTask$2.call(AsyncTask.java:305)
10-12 23:37:19.959  6513  6589 W System.err:    at java.util.concurrent.FutureTask.run(FutureTask.java:237)
10-12 23:37:19.959  6513  6589 W System.err:    at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:243)
10-12 23:37:19.960  6513  6589 W System.err:    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
10-12 23:37:19.960  6513  6589 W System.err:    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
10-12 23:37:19.960  6513  6589 W System.err:    at java.lang.Thread.run(Thread.java:761)
```
- [ ] No error message when clicking checkmark early
- [ ] Checks for fields missing (even URL can be empty, after which you can't open the repo again due to "unsupported repo type")
- [ ] On entering "asdasd" as URL
  - [ ] Some dialog is briefly displayed in the middle (what is that?)
  - [ ] "Unable to clone" Snackbar is displayed - all data needs to be checked first before trying to do anything
- [ ] Remove context from App, no need currently for it (pass it to `GitRepo` as it's done for others)
- [ ] Move git package to sync/ or repos/ (for now)
- [ ] Allow login with user/pass (could wait for release after the first one probably)

**Comment by timoc** [comment]
Is this available to play with?

**Comment by IvanMalison** [comment]
@timoc 
Here is a build of this pull request:
https://keybase.pub/imalison/orgzly-git-sync.apk

**Comment by timoc** [comment]
@IvanMalison got it working. So far so good. 

Did you know that gitlab can host private repos, and has an online edtitor that can render org files?

**Comment by IvanMalison** [comment]
@timoc no, but thats awesome. I might have to switch over from bitbucket. You might also be interested in this:

https://keybase.io/blog/encrypted-git-for-everyone

**Comment by IvanMalison** [comment]
Here are some more things that need to be fixed:

[] Merge commit is not done with provided git user
[] Merge commit string needs additional spaces in a few places

**Comment by timoc** [comment]
Using gitlab to change the files is proving difficult. especially if i use a branch.
Questions:
- Does the branching mechanism work?
- Can you explain a bit the git workflow you have implemented - local and remote?
Issues: 
"Clearing the Database " resets the repo information, can you add a 'clear content' or 'force resync' to do a  'git reset --hard'

**Comment by IvanMalison** [comment]
> Does the branching mechanism work?

Branching isn't really supported yet, except that you can specify which branch of the remote you would like to interact with as a repository setting.

> Can you explain a bit the git workflow you have implemented - local and remote?
> Issues:

I use Emacs to edit my org files, and I automatically synchronize the files with git using sparkleshare. I don't think theres much point to using org-mode unless you are using emacs.

> "Clearing the Database " resets the repo information, can you add a 'clear content' or 'force resync' to do a 'git reset --hard'

**Comment by timoc** [comment]
Hi,

Nothing much more to add yet. 

Is the apk link (https://keybase.pub/imalison/orgzly-git-sync.apk) up to date with these changes?

**Comment by IvanMalison** [comment]
@timoc yes

**Comment by nevenz** [comment]
Tried again, above comment updated. I have also separated all the nitpicking stuff that I'll know how to do myself.

**Comment by timoc** [comment]
A new binary too?

**Comment by f0086** [comment]
I've tried this branch and test it a little, but I am not quiet sure how the workflow of the git integration is intended. Should I already have a cloned the git repo on the device? I've getting the error "this is not a git repo" when I choose an empty folder for the git repo. Thought this is the folder in which the repo (i've specified on top) will be cloned. All in all, this branch feels a little bit rough and need a good portion of polish before going live (I've also had to fix that flavorDimensions thing). Is there still active development on this branch?

**Comment by IvanMalison** [comment]
@f0086 

> Should I already have a cloned the git repo on the device? I've getting the error "this is not a git repo" when I choose an empty folder for the git repo.

The way it currently works is that you need to choose a non-existent folder if you would like the clone to be performed by the orgzly app. If the folder exists, it assumes that you are indicating that you would like to use that folder as an existing repository. The reason that it has this behavior is that it would be bad (IMO) if orgzly clobbered an existing directory that might have content that is important to the user. I suppose that a check could be performed to see if the folder is empty, and in that case, move forward with the clone.

> All in all, this branch feels a little bit rough and need a good portion of polish before going live

Yeah, I'll concede that I didn't really make the UI work super smoothly, because this whole change ended up being a lot more work than I was anticipating, partly because this was my first foray into Android development. Once I got it to a state where it worked for me, I kinda lost motivation to give it (especially the user interface) that extra polish.

I assure you, however, that the actual git synchronization logic is incredibly sound, to the point where I would argue that it is much less likely to cause any sort of data loss that existing synchronization methods.

> Is there still active development on this branch?

I do have the intention to come back to this in the near future to finish it up, but I must concede that I didn't really enjoy this project (being that it's in Java), and the work that remains (mostly UI stuff), is especially tedious and uninteresting to me. I also have "had the intention" to come back to it for about 2 months now, but I just haven't been able to find the time yet.

**Comment by IvanMalison** [comment]
@codygman Yeah. I'd love it if this got merged and made a release.

Looks like theres a crap ton of conflicts though.

**Comment by nevenz** [comment]
> @codygman Yeah. I'd love it if this got merged and made a release.
> Looks like theres a crap ton of conflicts though.

I'll be focusing on new syncing methods for [v1.7](https://github.com/orgzly/orgzly-android/milestone/10), including this one. I haven't scheduled it for v1.7 yet, as I don't know right now how tricky will be to merge it, but I'll be having it in mind and working towards that (like modifying interface for `TwoWaySync` and adding repository type to stop relying on URLs to determine it).

**Comment by IvanMalison** [comment]
> like modifying interface for TwoWaySync and adding repository type to stop relying on URLs to determine it)

Yeah I think this is pretty crucial.

**Comment by IvanMalison** [comment]
@nevenz Did my best to resolve conflicts, but I haven't touched android studio in a while and can't seem to get things working right now.

The original history can be found here:
https://github.com/IvanMalison/orgzly-android/tree/gitSyncOriginal

**Comment by nevenz** [comment]
> @nevenz Did my best to resolve conflicts,

Thanks for taking the time to do this. I was wondering what to do with it as I started moving repository fragments to activities, which would make merging even more difficult.

It's merged to master now, so it doesn't fall behind again.

It's hidden under `IS_GIT_ENABLED` (set to `false` in `build.gradle`) until all issues are fixed. I've just fixed compilation issues and one crash for now.

There's no browser fragment anymore, so currently nothing happens on "Browse". I'll see to fix that and make remaining fragments (directory and now git) into activities.

I'll keep the above comment updated.

>  but I haven't touched android studio in a while and can't seem to get things working right now.

You might need the latest version if haven't tried it already.

**Comment by nevenz** [comment]
> It's hidden under IS_GIT_ENABLED (set to false in build.gradle)

It's `git.enabled` in [app.properties](https://github.com/orgzly/orgzly-android/blob/master/sample.app.properties#L23) now.

**Comment by bpiel** [comment]
Can I please send someone money to make this happen? If there were a kickstarter (or similar), I'd donate and help promote it.

**Comment by IvanMalison** [comment]
@bpiel you can always use the link I posted above to get a really old version that has working git synchronization:

https://keybase.pub/imalison/orgzly-git-sync.apk

I looked at trying to get stuff working again briefly when I did the merge a while back (Guess it was May 20th) and it seemed like a bit too much for me to take on at that moment.

**Comment by caleb-allen** [comment]
What is the status of this? What needs to get done?

I would be willing to pick this up, if I can find some free time. I'm an android dev and would be comfortable building a diff gui of some sort, or whatever needs doing. Mostly because I would really like this feature and Orgzly doesn't offer me much value until it's in.

**Comment by IvanMalison** [comment]
@caleb-allen see the first comment on https://github.com/orgzly/orgzly-android/pull/543#issue-278409449 . Would love it if you could finish this off. I'm not an android dev and it just became too painful for me to do the UI work.

**Comment by reyman** [comment]
Thanks @IvanMalison to work on that, i see you use keybase, is there any chance that this plugin work with encrypted keybase:// git capabilities ?

**Comment by reyman** [comment]
@IvanMalison Seems some people use another strategy with git-remote-gcrypt on termux :  https://github.com/keybase/kbfs/issues/1231

**Pull request merged**